hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-nodemanager/src/main/java/org/apache/hadoop/yarn/server/nodemanager/containermanager/linux/resources/CGroupsV2ResourceCalculator.java

```diff
@@ -0,0 +1,135 @@
+import java.util.stream.Stream;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.apache.commons.io.FileUtils;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.hadoop.classification.VisibleForTesting;
+import org.apache.hadoop.util.CpuTimeTracker;
+import org.apache.hadoop.yarn.exceptions.YarnException;
+
+/**
+ * A CGroupV2 file-system based Resource calculator without the process tree features.
+ *
+ * The feature only works if cluster runs in pure V2 version, because when we read the
+ * /proc/{pid}/cgroup file currently we can not handle multiple lines.
+ */
+public class CGroupsV2ResourceCalculator extends AbstractCGroupsResourceCalculator {
+  private static final Logger LOG = LoggerFactory.getLogger(CGroupsV2ResourceCalculator.class);
+  private final Map<String, String> stats = new ConcurrentHashMap<>();
+
+  @VisibleForTesting
+  String root = "/";
+
+  /**
+   * Create resource calculator for the container that has the specified pid.
+   * @param pid A pid from the cgroup or null for all containers
+   */
+  public CGroupsV2ResourceCalculator(String pid) {
+    super(pid);
+  }
+
+  @Override
+  public void initialize() throws YarnException {
+    cpuTimeTracker = new CpuTimeTracker(jiffyLengthMs);
+    cGroupsHandler = ResourceHandlerModule.getCGroupsHandler();
+  }
+
+  @Override
+  public long getCumulativeCpuTime() {
+    // https://docs.kernel.org/admin-guide/cgroup-v2.html#cpu-interface-files
+    return jiffyLengthMs < 0
+        ? UNAVAILABLE
+        : getStat("cpu.stat#usage_usec") * jiffyLengthMs;
+  }
+
+  @Override
+  public long getRssMemorySize(int olderThanAge) {
+    // https://docs.kernel.org/admin-guide/cgroup-v2.html#memory-interface-files
+    return 1 < olderThanAge
+        ? UNAVAILABLE
+        : getStat("memory.stat#anon");
+  }
+
+  @Override
+  public long getVirtualMemorySize(int olderThanAge) {
+    // https://docs.kernel.org/admin-guide/cgroup-v2.html#memory-interface-files
+    return 1 < olderThanAge
+        ? UNAVAILABLE
+        : getStat("memory.stat#vmalloc");
+  }
+
+  @Override
+  public void updateProcessTree() {
+    try (Stream<Path> cGroupFiles = Files.list(getCGroupPath())){
+      List<Path> statFiles = cGroupFiles
+          .filter(path -> path.toString().endsWith(".stat"))
+          .collect(Collectors.toList());
+      for (Path statFile : statFiles) {
+        String[] lines = fileToString(statFile).split(System.lineSeparator());
+        for (String line: lines) {
+          String[] parts = line.split(" ");
+          stats.put(statFile.getFileName() + "#" + parts[0], parts[1]);
+        }
+      }
+      cpuTimeTracker.updateElapsedJiffies(
+          BigInteger.valueOf(getStat("cpu.stat#usage_usec")), clock.getTime());
+      LOG.debug("The {} process has the following stat properties updated: {}", pid, stats);
+    } catch (Exception e) {
+      LOG.warn("Failed to read CGroupV2 stats for process: " + pid, e);
+    }
```

Comment:
For example: this is essentially the same method as `CGroupsResourceCalculator.updateProcessTree` just operates on different files. Basically read the jiffies, update the elapsed jiffies and then update the virtual/physical memory variables. We can edit the v1 method slightly but it would be good to keep one implementation only.